{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " adata = sc.read_h5ad('/projects/bioinformatics/DB/scRNAseq_parkinson/metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"your_data.h5ad\")  # replace with your file\n",
    "\n",
    "# 1. Normalize the data\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# 2. Identify highly variable genes (HVGs)\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n",
    "adata_hvg = adata[:, adata.var.highly_variable]\n",
    "\n",
    "# 3. Scale the data (zero mean, unit variance)\n",
    "sc.pp.scale(adata_hvg)\n",
    "\n",
    "# 4. Compute PCA\n",
    "sc.tl.pca(adata_hvg, svd_solver='arpack')\n",
    "\n",
    "# 5. Visualize PCA colored by Leiden clusters or cell type\n",
    "sc.pl.pca_scatter(adata_hvg, color=['cell_type'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "correct_census2025",
   "language": "python",
   "name": "cellxgene_census2025_new"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
